Zend/zend_compile.c

```diff
@@ -4712,6 +4712,171 @@ static void zend_compile_ns_call(znode *result, znode *name_node, zend_ast *args
 }
 /* }}} */
 
+static zend_op *zend_compile_rope_add(znode *result, uint32_t num, znode *elem_node);
+static zend_op *zend_compile_rope_add_ex(zend_op *opline, znode *result, uint32_t num, znode *elem_node);
+static void zend_compile_rope_finalize(znode *result, uint32_t j, zend_op *init_opline, zend_op *opline);
+
+static zend_result zend_compile_func_sprintf(znode *result, zend_ast_list *args) /* {{{ */
+{
+	if (args->children < 1) {
+		return FAILURE;
+	}
+
+	zend_eval_const_expr(&args->child[0]);
+	if (args->child[0]->kind != ZEND_AST_ZVAL) {
+		return FAILURE;
+	}
+
+	zval *format_string = zend_ast_get_zval(args->child[0]);
+	if (Z_TYPE_P(format_string) != IS_STRING) {
+		return FAILURE;
+	}
+	if (Z_STRLEN_P(format_string) >= 256) {
+		return FAILURE;
+	}
+
+	char *p;
+	char *end;
+	uint32_t string_placeholder_count;
+	
+	string_placeholder_count = 0;
+	p = Z_STRVAL_P(format_string);
+	end = p + Z_STRLEN_P(format_string);
+
+	for (;;) {
+		p = memchr(p, '%', end - p);
+		if (!p) {
+			break;
+		}
+
+		char *q = p + 1;
+		if (q == end) {
```

Comment:
I find the explicit boundary check easier to understand from a correctness perspective.

Given the complexity of preserving the semantics here, I preferred this initial version to be pretty explicit about what it does, making it easier to review, instead of trying to squeeze out the last bit of performance.

Reply:
Personally, I find redundant/conflicting checks more confusing, because it makes me question my understanding. ^^ But different minds think differently, so.